src/server.rs

```diff
@@ -1,62 +1,119 @@
-use futures::{future, executor, Async, Future, Stream};
+use bytes::Bytes;
+use futures::{Async, Poll, Future, Stream};
 use flate2::Compression;
-use flate2::read::GzEncoder;
+use flate2::write::GzEncoder;
 use hyper::{self, Body};
-// use hyper::header;
 use hyper::server::Service;
 use http::{self, header};
-use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_io::AsyncRead;
 
-// use ::MaybeCompressed;
+use std::io::{Write, Cursor};
 
-use std::io::{Read, Write, Cursor};
-use std::thread;
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Builder {
+    chunk_size: Option<usize>,
+    compression: Option<Compression>,
+}
 
-pub struct Gzip<T> {
+#[derive(Debug, Clone)]
+pub struct GzipChunked<T> {
     inner: T,
+    chunk_size: usize,
+    compression: Compression,
 }
 
-// fn map_response<B, F>(rsp: http::Response<B>, f: F)
-//     -> http::Response<MaybeCompressed<B>>
-// where
-//     F: Fn(B) -> MaybeCompressed<B>,
-//     B: Write,
-// {
-//     let status = rsp.status();
-//     let headers = rsp.headers().clone(); // TODO: remove clone.
-//     let body: B = *(rsp.body_ref().unwrap().clone());
-//     let body = f(body);
-//     http::Response::new()
-//         .with_status(status)
-//         .with_headers(headers)
-//         .with_body(body)
-// }
-
-impl<T> Gzip<T> {
+const MAX_CHUNK_SIZE: usize = 2048;
+
+struct ChunkingStream<B: AsyncRead> {
+    read: B,
+    chunksz: usize,
+}
+
+// impl Builder
+impl Builder {
+    pub fn chunk_size(mut self, size: usize) -> Self {
+        self.chunk_size = Some(size);
+        self
+    }
+
+    pub fn compression_level(mut self, compression: Compression) -> Self {
+        self.compression = Some(compression);
+        self
+    }
+
+    pub fn to_service<T>(&self, inner: T) -> GzipChunked<T> {
+        GzipChunked {
+            inner,
+            chunk_size: self.chunk_size.unwrap_or(MAX_CHUNK_SIZE),
+            compression: self.compression.unwrap_or_default(),
+        }
+    }
+}
+
+impl<B: AsyncRead> Stream for ChunkingStream<B> {
+    type Item = Bytes;
+    type Error = hyper::Error;
+
+     fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let mut buf = vec![0u8; self.chunksz];
```

Comment:
This will mean that every poll, a new vector is allocated. It probably would be better to save the buffer on this struct. And seeing as you're returning `Bytes` from this, it could just be that...

Reply:
Ah, good point!